dbld/rules

```diff
@@ -134,12 +134,19 @@ shell-%: setup
 images: $(foreach image,$(IMAGES), image-$(image))
 image: image-$(DEFAULT_IMAGE)
 image-%:
-	cd $(DBLD_DIR)/images/ && $(DOCKER) build --network=host -t balabit/syslog-ng-$* -f $*.dockerfile .
+	cd $(DBLD_DIR)/images/ && $(DOCKER) build --build-arg=OS_PLATFORM=$* --build-arg=COMMIT=$$(git rev-parse --short HEAD) --network=host -t balabit/syslog-ng-$* -f $*.dockerfile .
+
+check-image-%:
+	@image_commit=$$(docker inspect  --format '{{ index .Config.Labels "COMMIT"}}' balabit/syslog-ng-$*);		\
+	dbld_changed=$$(git diff $${image_commit:-NO_COMMIT_LABEL_IN_DBLD_IMAGE} -- dbld | wc -l);		\
```

Comment:
Although it "handles" the case when there is no commit label, but `| wc -l` is going to result in `0` thus no change, which is kinda dangerous.
My biggest issue with it that it kinda assumes that the commit is in fact in the repository. Which can be untrue in numerous case:
* because of github workflow one fork could miss commits from upstream other forks
* for example travis does shallow copy of repositories (so it could miss commit even if it is in the repository)

In those case it is going to report no change was found.

My alternative suggestion is to use the commit hash of the `dbld` image when the image is created, and simple compare those hash values if different there was probably a change. But even if it is a false report it is better than silently ignore a possible change.

You could use: `git rev-list --abbrev=7 --abbrev-commit -1 HEAD -- dbld`

Reply:
I actually prefer storing the GIT commit instead of an obscure object within the repository, as the COMMIT is usable for use-cases outside of this caching.

I don't see how the github workflow would cause missing commits. dockerhub will always build from master, so any branch forking off of master should get the commit that was used to generate the image (except if it reverted it, but still the commit would be available and the diff would be non-empty).

On the other hand shallow copies could cause problems, I agree. That could be mitigated by using a larger depth value (travis uses 50 by default but that can be customized). How is this handled by the `scripts/version.sh` script?

Reply:
What kind of other use-cases do you have in mind ?

If you have like the `original` and `fork_a` and `fork_b`, both `fork_a`. A docker image is built on `fork_a` with commit hash `forkAhash` and also a build made on `fork_b` via `forkBhash`. Let's also assume that both fork modifies `dbld` related functionality. Neither `fork_a` nor `fork_b` is going to have the same commit, as they are not integrated into the master.

The builder caches the `fork_a` (as it was the first), and now builds `fork_b`, which does not have `forkAhash` thus this code implies no change in the `dbld` using the old img. While in fact there was change!

(maybe I am missing some point, please correct me if so)

The `scripts/version.sh` handles in a way, that does not handle shallow copy and falls back to the `VERSION` file. (this also falls back to the same original behaviour)

Reply:
> What kind of other use-cases do you have in mind ?

having the COMMIT available will document exactly which commit was used to generate a specific image, that can possibly be corelated back to a jenkins/travis/dockerhub build as they all save the git commit for every build. 

It also gives us a sense of time (as commits do have dates associated with them), whereas if we only annotate using the object id of the directory we lose that. The dbld images should only depend on the contents of the "dbld" directory, still I feel that associating images with git IDs is more natural.

It could possibly make sense to use the dbld object id as an image version in docker (e.g instead of "latest"), which could allow us to use multiple images in our cache. But this would make referencing images less intuitive

> 
> If you have like the `original` and `fork_a` and `fork_b`, both `fork_a`. A docker image is built on `fork_a` with commit hash `forkAhash` and also a build made on `fork_b` via `forkBhash`. Let's also assume that both fork modifies `dbld` related functionality. Neither `fork_a` nor `fork_b` is going to have the same commit, as they are not integrated into the master.
> 
> The builder caches the `fork_a` (as it was the first), and now builds `fork_b`, which does not have `forkAhash` thus this code implies no change in the `dbld` using the old img. While in fact there was change!
> 
> (maybe I am missing some point, please correct me if so)
> 

I see your point now. Let me think a bit.

> The `scripts/version.sh` handles in a way, that does not handle shallow copy and falls back to the `VERSION` file. (this also falls back to the same original behaviour)

Reply:
I misunderstood your idea about the commit ID of the last dbld change. I thought you referred to the object id of the DBLD directory. But as far as I see directories do not even have that (only files have a content id, directories don't).

Reply:
You can ask for commit hash for a files, and for directories as well(if directory is tracked indirectly via tracking a file inside that directory).
The proposal was to query and use that has (for command see my original post).

Out of the original caching solution and the one I mentioned I would prefer mine. I am okay with different solution, it just have to detect the change in the above written cases.

Reply:
I understand that you can get a commit ID that last changed a specific directory. The question is in what way should we use it?

1) use it for the docker image ID (e.g. instead of `latest` we would have a series of docker images each tagged with the commit that last changed dbld). I have decided against this, as I think there's a lot of temporal state encapsulated in an image. Which means that one `dbld` revision is not the same as another one depending on the time the image was generated (because apt/pip/obs/etc changes). This means that if we used docker revisions, we would need to start maintaining those revisions, and regenerate them from time to time. The difficulty comes from knowing how far we need to go back maintaining these images and how to remove stale ones.

2) use it for a label within the specific image (e.g. we are using `latest`, but we'd have a DBLD_COMMIT label within the image). I don't see how this helps us over what I have in there currently. The change detection might be simpler (e.g. just compare two commit IDs), but we would need the same history anyway (to come up with our `dbld` commit id we need to use rev-list, that again needs history).

So my rationale has been for just using COMMIT
1) I don't want to maintain a series of dbld builds (e.g. latest only)
2) COMMIT is enough for change detection and it imposes the same requirements as DBLD_COMMIT would (e.g. we would need git history on the build machine that should go back to the latest image built by dockerhub).
3) I don't think dbld would change that often, but I expect dockerhub to build daily from master, and I would like to make it trivial which commit dockerhub was using to build an image.

With those in the latest revision, I think I've handled all cases where local history is not available (e.g. if the commit is not there I _always_ trigger a rebuild of the image). This is not a perfect caching decision, for that we would need the entire history, but should work in most of the cases, speeding up the builds considerably as long as the previous build image is available on the build host.

Reply:
> if the commit is not there I always trigger a rebuild of the image

That's the important thing I wanted, and did not care what metadata storage achieves it (only tried to provide one that I know would work, and never forcing to use that).
I'll check the code.

Reply:
First things first. I do not have a proposal yet, and I don't know a better solution. Sorry for that.
My concerns/how do I see it:

With image caching we introduced the plain old caching problem. We don't know what time we should refresh the cache.
- Currently we do not follow / mark the versions of the upstream dependencies.
- Using COMMIT instead of DBLD_COMMIT "just" increase the refreshing frequency, but still don't make any logical connection with the content of the image. (i.e.: a simple date check will do the same)

Totally agree with Bazsi's 1. point.
> I think there's a lot of temporal state encapsulated in an image. Which means that one dbld revision is not the same as another one depending on the time the image was generated (because apt/pip/obs/etc changes)

Only my conclusion differs. In theory: Two docker image should be identical if DBLD_COMMIT hasn't change, regardless of the master commit. And neither one of them is suitable to track the "temporal state" changes.